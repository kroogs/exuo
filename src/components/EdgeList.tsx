/*
 * Copyright Â© 2020 Ty Dira <ty@dira.dev>
 *
 * This file is part of Exuo.

 * Exuo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * Exuo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with Exuo.  If not, see <https://www.gnu.org/licenses/>.
 */

import React from 'react'
import {
  List,
  ListItem,
  ListItemText,
  ListItemSecondaryAction,
} from '@material-ui/core'
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles'
import { IAnyModelType, Instance } from 'mobx-state-tree'
import { useObserver } from 'mobx-react-lite'

import { Node } from 'graph'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    listSection: {},
    listItem: {
      padding: theme.spacing(2),
      paddingTop: theme.spacing(1),
      paddingBottom: theme.spacing(1),
    },
    itemText: {
      margin: 0,
      '& > .MuiListItemText-primary': {
        overflowX: 'hidden',
        whiteSpace: 'nowrap',
        textOverflow: 'ellipsis',
      },
    },
    group: { padding: 0 },
    groupHeader: {
      top: 0,
      position: 'sticky',
      fontWeight: 'bold',
      lineHeight: 2.5,
      backgroundColor: theme.palette.background.default,
    },
  }),
)

type SelectHandler = (
  event: React.SyntheticEvent,
  parent: Instance<typeof Node>,
  selected: Instance<typeof Node>,
) => void

interface EdgeListProps {
  node: Instance<IAnyModelType>
  tag: string
  onSelect?: SelectHandler
  className?: string
}

const EdgeList: React.FunctionComponent<EdgeListProps> = ({
  node,
  onSelect,
  tag,
  className,
}) => {
  const classes = useStyles()

  return useObserver(() => {
    return (
      <List aria-label="edge list" className={className}>
        {node.edgeMap
          .get(tag)
          ?.slice()
          .reverse()
          .map((item: Instance<typeof Node>) => (
            <ListItem
              button
              onClick={e => onSelect?.(e, node, item)}
              onDoubleClick={e => console.log('banana')}
              key={`${tag}-${item.id}`}
              className={classes.listItem}
            >
              <ListItemText
                primary={item.label ?? item.id}
                className={classes.itemText}
              />
              <ListItemSecondaryAction>99</ListItemSecondaryAction>
            </ListItem>
          ))}
      </List>
    )
  })
}

export default EdgeList
