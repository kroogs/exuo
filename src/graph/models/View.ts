/*
 * Copyright Â© 2020 Ty Dira <ty@dira.dev>

 * This file is part of Exuo.

 * Exuo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * Exuo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with Exuo.  If not, see <https://www.gnu.org/licenses/>.
 */

import { types, Instance } from 'mobx-state-tree'

import { nodeFactory } from 'graph'

const ViewerConfigItem = types
  .model('ViewerConfigItem', {
    expanded: types.optional(types.boolean, false),
  })
  .actions(self => ({
    setExpanded(value: boolean) {
      self.expanded = value
    },
  }))

const ViewerConfigBase = types
  .model('ViewerConfig', {
    items: types.map(types.map(ViewerConfigItem)),
  })

  .views(self => ({
    getItem(
      parentId: string,
      childId: string,
    ): Instance<typeof ViewerConfigItem> | void {
      if (self.items.size) {
        const parent = self.items.get(parentId)
        if (parent && parent.size) {
          return parent.get(childId)
        }
      }
    },
  }))

  .actions(self => ({
    initItem(
      parentId: string,
      childId: string,
    ): Instance<typeof ViewerConfigItem> | void {
      const item = self.getItem(parentId, childId)

      if (!item) {
        const parent = self.items.get(parentId)
        const item = ViewerConfigItem.create()

        if (parent) {
          parent.set(childId, item)
        } else {
          self.items.set(parentId, {
            [childId]: item,
          })
        }
      }

      return self.getItem(parentId, childId)
    },
  }))

export const ViewerConfig = nodeFactory(ViewerConfigBase)
