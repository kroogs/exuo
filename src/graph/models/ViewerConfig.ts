/*
 * Copyright Â© 2020 Ty Dira <ty@dira.dev>

 * This file is part of Exuo.

 * Exuo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * Exuo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with Exuo.  If not, see <https://www.gnu.org/licenses/>.
 */

import { types, Instance } from 'mobx-state-tree'

import { nodeFactory } from 'graph'

const ViewerConfigItem = types
  .model('ViewerConfigItem', {
    expanded: types.optional(types.boolean, false),
  })
  .actions(self => ({
    setExpanded(value: boolean) {
      self.expanded = value
    },
  }))

const ViewerConfigBase = types
  .model('ViewerConfig', {
    items: types.map(types.map(ViewerConfigItem)),
  })
  .actions(self => ({
    getItem(
      parentId: string,
      childId: string,
      doCreate?: boolean,
    ): Instance<typeof ViewerConfigItem> | void {
      let item = self.items.get(parentId)?.get(childId)

      if (!item && doCreate) {
        item = ViewerConfigItem.create()

        self.items.set(parentId, {
          [childId]: item,
        })
      }

      return item
    },
  }))

export const ViewerConfig = nodeFactory(ViewerConfigBase)
