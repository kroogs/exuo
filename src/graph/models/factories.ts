/*
 * Copyright Â© 2020 Ty Dira <ty@dira.dev>

 * This file is part of Exuo.

 * Exuo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * Exuo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with Exuo.  If not, see <https://www.gnu.org/licenses/>.
 */

import {
  IAnyModelType,
  IAnyType,
  Instance,
  isIdentifierType,
  SnapshotIn,
  isModelType,
  types,
} from 'mobx-state-tree'
import { v4 as uuid } from 'uuid'

export type ModelOrUnion = IAnyModelType | IAnyType
export type EdgeResolver = () => ModelOrUnion
export type ModelTable = Record<string, IAnyModelType>

export const edgeMapFactory = (getEdgeType: EdgeResolver): IAnyModelType =>
  types
    .model('EdgeMap', {
      edgeMap: types.map(
        types.array(
          types.safeReference(types.late(getEdgeType), {
            acceptsUndefined: false,
          }),
        ),
      ),
    })
    .actions(self => ({
      hasEdge(tag: string, target?: Instance<IAnyModelType>): boolean {
        if (target) {
          return Boolean(self.edgeMap.get(tag)?.includes(target))
        } else {
          return Boolean(self.edgeMap.get(tag)?.length)
        }
      },
    }))
    .actions(self => ({
      addEdge(tag: string, target: Instance<IAnyModelType>): void {
        if (!self.edgeMap.has(tag)) {
          self.edgeMap.set(tag, [])
        }

        self.edgeMap.get(tag)?.push(target)
      },

      removeEdge(tag: string, target: Instance<IAnyModelType>): void {
        if (!self.edgeMap.get(tag)?.remove(target)) {
          throw Error(`Node '${self}' has no '${tag}' edge for '${target}'`)
        }
      },
    }))
    .views(self => ({
      getEdgeTag(tag: string): void | Array<Instance<IAnyModelType>> {
        return self.edgeMap.get(tag)
      },
    }))

export const nodeFactory = (
  models: IAnyModelType | Array<IAnyModelType> = [],
): IAnyModelType =>
  types
    .compose(
      // @ts-ignore
      ...[
        types.model({
          id: types.identifier,
        }),
      ].concat(models),
    )
    .named('Node')

interface GraphFactoryOptions {
  getId?: () => string
}

export const graphFactory = (
  nodeModels: ModelTable,
  options: GraphFactoryOptions = {},
): IAnyModelType =>
  types
    .model(
      'Graph',
      Object.fromEntries(
        Object.keys(nodeModels).map(key => {
          const model = nodeModels[key]

          if (isModelType(model) && isIdentifierType(model.properties.id)) {
            return [key, types.map(nodeModels[key])]
          }

          // Can't add an identifier because the model may be referenced
          throw Error(`Model '${key}' requires 'id' identifier`)
        }),
      ),
    )
    .actions(self => ({
      createNode(
        modelName = 'Node',
        props: SnapshotIn<IAnyModelType> = {},
      ): Instance<IAnyModelType> {
        if (!nodeModels[modelName]) {
          throw Error(`No model named '${modelName}'`)
        }

        const node = nodeModels[modelName].create({
          id: options.getId?.() ?? uuid(),
          ...props,
        })

        self[modelName].put(node)

        return node
      },
    }))

export const NodeBase = nodeFactory(edgeMapFactory(() => NodeBase))
export const GraphBase = graphFactory({ Node: NodeBase })
