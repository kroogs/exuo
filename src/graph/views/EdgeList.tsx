/*
 * Copyright Â© 2020 Ty Dira <ty@dira.dev>

 * This file is part of Exuo.

 * Exuo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * Exuo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with Exuo.  If not, see <https://www.gnu.org/licenses/>.
 */

import React from 'react'
import { List } from '@material-ui/core'
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles'
import { IAnyModelType, Instance } from 'mobx-state-tree'

import { Node, useGraph } from 'graph'
import NodeListItem from './NodeListItem'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    list: {},
  }),
)

interface EdgeListProps {
  node: Instance<IAnyModelType>
  edgeTag: string
  className?: string
}

export const EdgeList: React.FunctionComponent<EdgeListProps> = ({
  node,
  edgeTag,
  className,
}) => {
  const classes = useStyles()
  return useGraph(graph => (
    <List aria-label="edge list" className={classes.list}>
      {node.edgeMap
        .get(edgeTag)
        ?.slice()
        .reverse()
        .map((item: Instance<typeof Node>) => (
          <NodeListItem node={item} key={`${edgeTag}-${item.id}`} />
        ))}
    </List>
  ))
}
