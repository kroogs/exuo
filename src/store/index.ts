/*
 * Copyright Â© 2020 Ty Dira <ty@dira.dev>

 * This file is part of Exuo.

 * Exuo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * Exuo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with Exuo.  If not, see <https://www.gnu.org/licenses/>.
 */

export * from './StoreProvider'
export * from './useStore'
export * from './persist'

export * from './models/Store'

export * from './components/InstanceViewer'
export * from './components/PropertyEditor'
