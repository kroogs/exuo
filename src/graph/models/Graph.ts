/*
 * Copyright Â© 2020 Ty Dira <ty@dira.dev>

 * This file is part of Exuo.

 * Exuo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * Exuo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with Exuo.  If not, see <https://www.gnu.org/licenses/>.
 */

import { Instance, SnapshotIn, IAnyModelType } from 'mobx-state-tree'

import { persist } from 'store/persist'

import { graphFactory } from './factories'
import { Node, Config, Note } from './Node'

export const Graph = graphFactory({ Node, Config, Note })
  // Modes
  .actions(self => ({
    setActiveMode(mode: string) {
      const activeModes = self.Config.get('system').get('activeModes')
      if (activeModes && !activeModes?.includes(mode)) {
        activeModes.push(mode)
      }
    },

    unsetActiveMode(mode: string) {
      const activeModes = self.Config.get('system').get('activeModes')
      if (activeModes?.includes(mode)) {
        activeModes.remove(mode)
      }
    },

    toggleActiveMode(mode: string) {
      const activeModes = self.Config.get('system').get('activeModes')

      if (!activeModes) {
        return
      }

      if (activeModes.includes(mode)) {
        activeModes.remove(mode)

        // TODO mode logic doesn't belong here
        if (mode === 'select') {
          self.clearSelectedNodes()
        }
      } else {
        activeModes.push(mode)
      }
    },
  }))

  // Selection
  .actions(self => ({
    clearSelectedNodes() {
      self.Config.get('system').get('selectedNodes')?.clear()
    },

    deleteSelectedNodes() {
      const selectedNodes = self.Config.get('system')?.get('selectedNodes')
      selectedNodes.forEach((id: string) => self.Node.delete(id))
      self.clearSelectedNodes()
      if (self.selectedNodes.length === 0) {
        self.unsetActiveMode('select')
      }
    },

    toggleSelectNode(node: Instance<typeof Node>) {
      const selectedNodes = self.Config.get('system').get('selectedNodes')
      if (selectedNodes?.includes(node.id)) {
        selectedNodes?.remove(node.id)
      } else {
        selectedNodes?.push(node.id)
      }

      if (selectedNodes.length === 0) {
        self.toggleActiveMode('select')
      }
    },

    // selectNode(node: Instance<typeof Node>) {

    // },

    // deselectNode(node: Instance<typeof Node>) {

    // },

    selectAllChildNodes(node: Instance<typeof Node>) {
      node.edgeMap
        .get('child')
        .forEach((child: Instance<typeof Node>) => console.log('child'))
    },

    // deselectAllChildNodes(node: Instance<typeof Node>) {

    // },
  }))

  .actions(self => ({
    afterCreate() {
      persist(self).then(() => {
        if (!self.Config.has('system')) {
          const root = self.createNode('Node', { label: 'Exuo' })
          self.createNode('Config', {
            id: 'system',
            items: {
              rootNodeId: root.id,
              activeModes: [],
              selectedNodes: [],
            },
          })
          self.createNode('Config', {
            id: 'user',
            items: {
              global: {
                dividers: false,
              },
              lists: {
                checkbox: false,
                dividers: false,
                showChildCount: true,
                showEdgeChips: false,
              },
            },
          })
        }
      })
    },

    createChild<T extends IAnyModelType>(
      node: Instance<typeof Node>,
      edgeProps: SnapshotIn<T>,
      edgeType = 'Node',
    ): Instance<typeof Node> {
      const child = self.createNode(edgeType, edgeProps)

      node.addEdge('child', child)
      child.addEdge('parent', node)

      return child
    },

    deleteNode(node: Instance<typeof Node>) {
      self.Node.delete(node.id)
    },

    toggleGlobalDividerSetting() {
      const value = self.Config.get('user').get('global')?.get('dividers')
      self.Config.get('user').get('global')?.set('dividers', !value)
    },

    toggleListCheckboxSetting() {
      const value = self.Config.get('user').get('lists')?.get('checkbox')
      self.Config.get('user').get('lists')?.set('checkbox', !value)
    },
  }))
  .views(self => ({
    get rootNode() {
      const config = self.Config.get('system')
      return config && self.Node.get(config.get('rootNodeId'))
    },

    get activeModes() {
      const config = self.Config.get('system')
      return config?.get('activeModes')
    },

    get selectedNodes() {
      return self.Config.get('system')?.get('selectedNodes')
    },

    get listCheckboxSetting() {
      return self.Config.get('user')?.get('lists')?.get('checkbox')
    },
  }))
