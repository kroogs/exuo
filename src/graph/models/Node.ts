/*
 * Copyright Â© 2020 Ty Dira <ty@dira.dev>
 *
 * This file is part of Exuo.

 * Exuo is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.

 * Exuo is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.

 * You should have received a copy of the GNU Affero General Public License
 * along with Exuo.  If not, see <https://www.gnu.org/licenses/>.
 */

import { types, IAnyType } from 'mobx-state-tree'

import { nodeFactory, edgeMapFactory } from 'graph/factories'
import * as models from 'graph/models'

export const Config = nodeFactory(models.Config)
export const Node = nodeFactory([
  models.Log,
  edgeMapFactory(() =>
    types.union(
      types.late((): IAnyType => Node),
      Config,
    ),
  ),
])
  .props({
    label: types.maybe(types.string),
  })
  .actions(self => ({
    afterCreate() {
      self.logEvent('create')
    },
  }))
  .views(self => ({
    get childCount(): number {
      return self.edgeMap.get('child')?.length ?? 0
    },

    get createDate(): Date {
      return self.events[0]?.date
    },
  }))
