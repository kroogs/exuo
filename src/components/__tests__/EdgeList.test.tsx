/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 * Copyright Â© 2020 Ty Dira */

import React from 'react'
import { mount } from 'enzyme'

import EdgeList from '../EdgeList'
import { graphFactory, Node } from 'store/models'

let nextId = 0
const makeId = (): string => String(++nextId)

const Graph = graphFactory({ Node }, { makeId })

test('Renders a list of associated nodes', () => {
  const graph = Graph.create()
  const first = graph.createNode()
  const last = graph.createNode()

  let prev = null

  for (let i = 0; i < 5; i++) {
    const node = graph.createNode()

    node.addEdge('last', last)
    node.addEdge('first', first)
    first.addEdge('child', node)

    if (prev) {
      prev.addEdge('next', node)
      node.addEdge('prev', prev)
    }

    prev = node
  }

  expect(mount(<EdgeList node={first} />)).toMatchSnapshot()
})
