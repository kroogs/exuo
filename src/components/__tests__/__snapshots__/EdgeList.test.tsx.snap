// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders a list of associated nodes 1`] = `
<EdgeList
  node={
    Object {
      "edgeMap": Object {
        "child": Array [
          "3",
          "4",
          "5",
          "6",
          "7",
        ],
      },
      "id": "1",
      "label": undefined,
    }
  }
>
  <WithStyles(ForwardRef(List))
    aria-label="edge list"
  >
    <ForwardRef(List)
      aria-label="edge list"
      classes={
        Object {
          "dense": "MuiList-dense",
          "padding": "MuiList-padding",
          "root": "MuiList-root",
          "subheader": "MuiList-subheader",
        }
      }
    >
      <ul
        aria-label="edge list"
        className="MuiList-root MuiList-padding"
      />
    </ForwardRef(List)>
  </WithStyles(ForwardRef(List))>
</EdgeList>
`;
